### PR TITLE
Arrows on links to show direction

### LINKED ISSUE #14: Hovering a node does not produce very informative visuals.
## Describe the bug

Currently when we hover on a node it highlights all the links which have atleast 1 node as 
1. Hovered node
2. Nodes which can be reached with hovered node as source node
3. Nodes which can be reached from nodes in step 2 as source nodes and so on..

Hovering nodes highlight the specified links. But it is very difficult for a viewer to guess the logic happening behind the scene. Hence it is not soo informative for viewers. (Even I was not able to guess the logic without reading the code). See the screenshots section. 
The more informative logic will be to highlight the source and destination nodes (that are the neighbouring nodes) of the hovered node with different styles for source and destination. For example with different colors. 
## To Reproduce

1. Go to [website](http://dataviz.creativecommons.engineering/)
2. Zoom the website to have a clear view of nodes and links.
3. Hover at different nodes (atleast 7-8 nodes)
4. Can you guess the logic behind the hover-node event ?

## Expected behavior

We should show links where hovered link is source node and where hovered link is destination node with different styles, so that the user should not have to view the direction of link from the link text. Hence making it more informative.

## Screenshots

**Different hovered nodes**

Almost all links are highlighted
ice_screenshot_20200227-101618

Only 1 link is highlighted
ice_screenshot_20200227-101838

## Desktop (please complete the following information)
 - OS: Windows 10
 - Browser: Chrome
 - Version: 80

### PR DESCRIPTION
**Pull request by BharatNischal**
Fixes #14 

## Description

Added arrows to the links to show direction of the links. The size of arrow has been selected such that the arrow is visible even in low zoom.
When the nodes are not hovered/highlighted main focus is given on text label and when the node is hovered/highlighted the main focus is given to arrow.

## Screenshots
**Without hovering**
ice_screenshot_20200320-111929

**On hovering**
ice_screenshot_20200320-112005

## Checklist
- [x] My pull request has a descriptive title (not a vague title like `Update
  index.md`).
- [x] My pull request targets the `master` branch of the repository.
- [x] My commit messages follow [best practices][best_practices].
- [ ] I added tests for the changes I made (if applicable).
- [ ] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no
  visible errors.

[best_practices]:https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53

## Developer Certificate of Origin
```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

### CODE REVIEW

**Comment by BharatNischal** [comment]
@soccerdroid  @mathemancer  Please Review this PR

**Comment by soccerdroid** [comment]
Is there any workaround for the arrows covering the edges descriptions?

**Comment by BharatNischal** [comment]
@soccerdroid  I have tried to fix this issue by replacing the '<' and '>' character in the link text with the link arrow and by adding some spaces in between the source and destination name with the help of some mathematics.
Its working fine for links having large font-size but the arrows are coinciding with the text in case of small fontsize. (Please have a look at the screenshots). So do you want me to stay with this approach and fix this issue for smaller font-size (probabily just making the arrow size as a function of font-size) or do I consider some other solution? For example changing the color combinations so that the link-text is visible even after hovering or highlighting of link.
Please give me your opinion.

**For normal font-size**
ice_screenshot_20200417-140335

**For smaller font-size**
ice_screenshot_20200417-140231


> Is there any workaround for the arrows covering the edges descriptions?

**Comment by soccerdroid** [comment]
> @soccerdroid I have tried to fix this issue by replacing the '<' and '>' character in the link text with the link arrow and by adding some spaces in between the source and destination name with the help of some mathematics.
> Its working fine for links having large font-size but the arrows are coinciding with the text in case of small fontsize. (Please have a look at the screenshots). So do you want me to stay with this approach and fix this issue for smaller font-size (probabily just making the arrow size as a function of font-size) or do I consider some other solution? For example changing the color combinations so that the link-text is visible even after hovering or highlighting of link.
> Please give me your opinion.
> 
> **For normal font-size**
> ice_screenshot_20200417-140335
> 
> **For smaller font-size**
> ice_screenshot_20200417-140231
> 
> > Is there any workaround for the arrows covering the edges descriptions?

Let's try with the color combination you mention, you can attach the screenshots here as a comment as well.

**Comment by soccerdroid** [comment]
Hi @BharatNischal ! Did you try the color combination?

**Comment by BharatNischal** [comment]
> Hi @BharatNischal ! Did you try the color combination?

@soccerdroid  Soo sorry for the delay. I was a little busy but I will finish this within 2 days and will get back to you.

**Comment by BharatNischal** [comment]
@soccerdroid  I tried with color combination but it did not worked. No matter what color I choose the arrows are always above the text. So I tried this new solution where I translate the arrows when they are highlighted. please see the screenshots.
Or we can just place the text above the edge at the start too.
ice_screenshot_20200507-125438

ice_screenshot_20200507-125536
what you say?

**Comment by soccerdroid** [comment]
Is there any way we can move the arrow as close as possible to the node? That could be one workaround. Another one could be to set a smaller arrow size for shorter edges (though maybe this is more complex, not sure if this is possible with the current library). How do the arrows look like for shorter edges with your last solution?

**Comment by subhamX** [comment]
Hi @soccerdroid and @BharatNischal. I think we can use **linkDirectionalArrowRelPos** method provided by force-graph and set its value around **1**, to keep the arrow in close proximity to the node.

Please have a look at the snapshot (using 0.99).

linkArrowPos

**Comment by soccerdroid** [comment]
Thanks @subhamX , I think that could work. What do you say @BharatNischal ?

**Comment by BharatNischal** [comment]
> Thanks @subhamX , I think that could work. What do you say @BharatNischal ?

Yes thats a good solution. Thanks @subhamX  for the help.
Actually I was taking a break from coding for a week. I will fix the issue soon

**Comment by BharatNischal** [comment]
It seems that this PR is having a conflict because of some previously merged PRs . How can I resolve this?

**Comment by subhamX** [comment]
@BharatNischal It seems that your branch `BharatNischal:directed-Edges` is not synchronized with `creativecommons:master`. I suggest you to follow these steps to synchronize and resolve conflicts. 

Make sure you are on `directed-Edges` branch.

```bash
# Add upstream URL
git remote add upstream git@github.com:creativecommons/cccatalog-dataviz.git
# Fetch Information
git fetch upstream
# Rebase with upstream/master i.e creativecommons:master
git rebase upstream/master
```
After these, you will be prompted with Merge Conflict. Open your text editor and perform necessary changes which fix the conflict. 
```bash
# Add the index.html file
git add index.html
# Continue the rebase
git rebase --continue
# Push your changes to origin/directed-Edges
git push origin/directed-Edges
```

Hope this helps!

**Comment by BharatNischal** [comment]
@subhamX thanks for the help. Now the merge issue has resolved but it now shows 3 extra commits. Is it ok?

File: GSoC2019/src/index.html [hunk 1]
```diff
@@ -172,11 +179,15 @@
...
       .nodeCanvasObjectMode(() => 'replace')
-      .zoom(0.05)
-
+      .zoom(currentZoomLevel)
       //draw source and targets on the links
+      .linkDirectionalArrowLength(5)
+      .linkDirectionalArrowRelPos(0.99)
       .linkCanvasObjectMode(() => 'after')
+      .linkDirectionalArrowLength(5)
+      .linkDirectionalArrowRelPos(0.99)
```
Comment (soccerdroid) [inline GSoC2019/src/index.html:190]: These lines are duplicated

**Review by soccerdroid — CHANGES_REQUESTED** [review]

**Comment by soccerdroid** [comment]
> @subhamX thanks for the help. Now the merge issue has resolved but it now shows 3 extra commits. Is it ok?

I think it isn't. Correct me if I am wrong, but you modified only one js file, right?

**Comment by BharatNischal** [comment]
> > @subhamX thanks for the help. Now the merge issue has resolved but it now shows 3 extra commits. Is it ok?
> 
> I think it isn't. Correct me if I am wrong, but you modified only one js file, right?

Yes I just updated index.html
How can I remove the extra commits?

**Comment by subhamX** [comment]
@BharatNischal,  the steps I had mentioned earlier for rebasing shouldn't create any extra merge commits or any commits since rebase should have taken fast-forward route. 

I think that `632157c` commit is redundant and was created after the rebase operation. [Duplicacy](https://github.com/creativecommons/cccatalog-dataviz/pull/20#discussion_r433510391) is because of this commit. You can consider removing it since all changes are already present in previous commits.

Another small thing. Merge commits should arise if you pull the branch (git pull) and there are some changes in the server which are not present locally. `f3b6c90` and `d2a1103` are because of this. In case you wish to reduce the number of commits consider squashing them.

**Comment by soccerdroid** [comment]
Another option, and because it got kind of confusing this rebasing/merging operation, is that you create another branch from the master branch, and add your modified file here. I'm suggesting this because the changes are not long. Is this easier for you @BharatNischal ?

**Comment by BharatNischal** [comment]
> Another option, and because it got kind of confusing this rebasing/merging operation, is that you create another branch from the master branch, and add your modified file here. I'm suggesting this because the changes are not long. Is this easier for you @BharatNischal ?

Ok sure it will be relatively easy to create a new branch

**Comment by BharatNischal** [comment]
https://github.com/creativecommons/cccatalog-dataviz/pull/30#issue-428428436
This is the link to new PR

**Comment by soccerdroid** [comment]
Closing this as #30 has been accepted and merged. Thanks @BharatNischal !

**Pull request closed without merging**